ddclient.in

```diff
@@ -5170,6 +5185,113 @@ sub nic_dinahosting_update {
+
+Available configuration variables:
+  * password: The Gandi API key. If you don’t have one yet, you can generate
+              your production API key from the API Key Page (in the Security
+              section).
+  * ttl:      The time-to-live value associated with the updated DNS record.
+  * zone:     The DNS zone to be updated.
+
+Example ${program}.conf file entries:
+  ## Single host update.
+  protocol=gandi,                           \\
+  zone=example.com,                         \\
+  password=my-gandi-api-key,                \\
+  ttl=3600                                  \\
+  host.example.com
+
+  ## Multiple host update.
+  protocol=gandi,                           \\
+  zone=example.com,                         \\
+  password=my-gandi-api-key,                \\
+  ttl=3600                                  \\
+  hosta.example.com,hostb.sub.example.com
+EoEXAMPLE
+}
+
+######################################################################
+## nic_gandi_update
+######################################################################
+sub nic_gandi_update {
+    debug("\nnic_gandi_update -------------------");
+
+    # Update each set configured host.
+    foreach my $h (@_) {
+        my $ip = delete $config{$h}{'wantip'};
+        (my $hostname = $h) =~ s/\.\Q$config{$h}{zone}\E$//;
+
+        info("%s -- Setting IP address to %s.", $h, $ip);
+        verbose("UPDATE:", "updating %s", $h);
+
+        my $headers;
+        $headers  = "Content-Type: application/json\n";
+        $headers .= "Authorization: Apikey $config{$h}{'password'}\n";
+
+        my $data = encode_json({
+            items => [{
+                rrset_type   => is_ipv6($ip) ? "AAAA" : "A",
+                rrset_ttl    => $config{$h}{'ttl'},
+                rrset_values => [$ip],
+            }]
+        });
+
+        my $url = "https://$config{$h}{'server'}/$config{$h}{'script'}/livedns/domains/$config{$h}{'zone'}/records/$hostname";
+        my $reply = geturl({
+            proxy   => opt('proxy'),
+            url     => $url,
+            headers => $headers,
+            method  => 'PUT',
+            data    => $data,
+        });
+        unless ($reply) {
+            failed("%s -- Could not connect to %s.", $h, $config{$h}{'server'});
+            next;
+        }
+        next if !header_ok($h, $reply);
+        # Strip reply header.
+        $reply =~ s/^.*?\n\n//s;
+
+        my $response = eval { decode_json($reply) };
+        if (!defined $response) {
+            failed("%s -- Unexpected service response.", $h);
+            $config{$h}{'status'} = "bad";
+
+            next;
+        }
+
+        if (defined $response->{status} && $response->{status} eq "error") {
+            my @errors;
+            for my $err (@{$response->{errors}}) {
+                push(@errors, $err->{description});
+            }
+            failed("%s -- %s.", $h, join(", ", @errors));
+            $config{$h}{'status'} = "bad";
+
+            next;
+        }
```

Comment:
I don't see `status`, `errors`, or `description` in the API docs. It looks like it returns 201 if successful and an HTTP 4xx response on error, so execution would never reach this point on error due to the `header_ok` check.

Reply:
It is not documented but the API actually returns a `400 Bad Request` with `{"status": "error", "errors": [{"location": "body", "name": "items", "description": "'toto': must be an IP address (param: {'rrset_values': [u'toto'], 'rrset_ttl': 10800, u'rrset_name': u'test', 'rrset_type': u'A'})"}]}`.

I've filled in a bug report so they can include it into the documentation.